docs/Security_Enhancement_For_AI_Models.md

```diff
@@ -0,0 +1,246 @@
+
+A simple model server includes a interface to get the model information, such as id, framework, model URL, KBS URL etc. here is an example in CNAP:
+
+- Request
+
+```Python
+# HTTP Headers
+headers = {
+            "Accept": "application/json",
+            "Content-Type": "application/json"
+}
+# Request URL
+req_url = http://{model_info_url}/{model_id}
+```
+
+- Response
+
+```JSON
+{
+    "id": "c8b019e0-f4d8-4831-8936-f7f64ad99509",
+    "framework": "tensorflow",
+    "target": "object-detection",
+    "url": "http://{model_server_url}/tensorflow/ssdmobilenet_v10.pb.enc",
+    "name": "ssdmobilenet",
+    "version": "1.0",
+    "dtype": "int8",
+    "encrypted": true,
+    "kbs": "simple_kbs",
+    "kbs_url": "https://{key_brocker_server_url}",
+    "key_id": "88b3a4b3-ba4b-4e2b-b00e-bc32ca3d56be"
+}
+```
+
+A model provider should be implemented to get models information and download the model from model URL, if the model is encrypted, then invoke the key broker client to get the key from the KBS and decrypt the model.
+
+### 2.2 Key Broker Service
+
+An [abstract class](https://github.com/intel/cloud-native-ai-pipeline/blob/main/cnap/core/keybroker.py) `KeyBrokerClientBase` was introduced in CNAP project, it has an abstract method `get_key` to get a key from Key Broker Service (KBS). For a new KBS, a key broker client should be implemented to connect to KBS and get the key. `SimpleKeyBrokerClient` is an example to get a key from a simple KBS.
+
+An example flow of a simple KBS:
+
+- Accept a quote and a public key from client.
+- Verify the quote and do attestation, return if verify failed.
+- Get the user key from key management server (KMS), generate a symmetric wrapping
+    key (SWK) to encrypt the user key (wrapped_key).
+- Encrypt the SWK by the public key from client (wrapped_swk).
+
+For a key broker client, here is an example flow to get a key from KBS:
+
+- Generate 2048 bit RSA key pair (a public key and a private key).
+- Encode the public key to base64 for transferring (user_data).
+- Get quote in the TEE with the hash of the public key for measurement (quote).
+- Request wrapped_key and wrapped_swk from KBS with quote and user_data.
+- Decrypt the user key by the SWK.
+
+An example interface to communicate to the KBS:
+
+- Request
+
+```Python
+# HTTP Headers
+headers = {
+    "Accept": "application/json",
+    "Content-Type": "application/json",
+    "Attestation-Type": "TDX"
+}
+# HTTP Body
+req_body = {
+    "quote": quote,
+    "user_data": user_data
+}
+```
+
+- Response
+
+```JSON
+{
+    "wrapped_key":"",
+    "wrapped_swk":""
+}
+```
+
+### 2.3 Attestation
+
+CCNP is an easy way to get the quote for attestation, before getting the quote, a RSA key pair can be genereted and the public key can be measured as the user data input to get the quote.
```

Comment:
CCNP provides an easy way to get the quote for attestation. Before getting the quote, a RSA key pair need to be generated for wrapping purpose and the public key will be measured as the user data input to the quote fetching process.

Reply:
Done.